### PR TITLE
transport socket: api and implementation for startTls transport socket

### PR DESCRIPTION
**Pull request by cpakulski**
Commit Message:
PR defines configuration API and implementation for StartTls transport socket. StartTls is programmable transport socket which starts in clear-text but may programatically be switched to TLS.

Original discussion about use cases is here: https://github.com/envoyproxy/envoy/issues/9577

Design document is here: https://docs.google.com/document/d/1ajrQpOEuup04V95xDDY8hSigW4B-J3iuhwn8QQEl-q4/edit?usp=sharing

Risk Level: Low.
Testing: Yes. Added integration tests.

Docs Changes: Some doc sections are automatically produced, but they contain minimum info.

Release Notes: No.

### CODE REVIEW

File: source/extensions/transport_sockets/starttls/BUILD [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+licenses(["notice"])  # Apache 2
+
+# StartTls transport socket.
+
+envoy_extension_package()
+
+envoy_cc_extension(
+    name = "config",
+    srcs = [
```
Comment (asraa) [inline source/extensions/transport_sockets/starttls/BUILD:15]: just a drive-by: these can be omitted if they're empty

File: api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+option java_package = "io.envoyproxy.envoy.extensions.transport_sockets.starttls.v3";
+option java_outer_classname = "StarttlsProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: StartTls]
+// [#extension: envoy.transport_sockets.starttls]
+
+// StartTls transport socket addresses situations when a protocol starts in clear-text and
+// negotiates in-band switch to TLS. StartTls transport socket is protocol agnostic and requires
```
Comment (mattklein123) [inline api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto:23]:
```suggestion
// negotiates an in-band switch to TLS. StartTls transport socket is protocol agnostic and requires
```

File: api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+// a network filter which understands protocol exchange and state machine to signal to StartTls
```
Comment (mattklein123) [inline api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto:24]:
```suggestion
// a network filter which understands protocol exchange and a state machine to signal to the StartTls
```

File: api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// transport socket when switch to TLS is required.
```
Comment (mattklein123) [inline api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto:25]:
```suggestion
// transport socket when a switch to TLS is required.
```

File: api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+// [#not-implemented-hide:] Configuration for StartTls transport socket.
+// StartTls transport socket wraps two sockets:
+// - raw_buffer socket which is used at the beginning of the session
+// - TLS socket used when a protocol negotiates a switch to encrypted traffic.
+message StartTlsConfig {
+  // Configuration for clear-text socket used at the beginning of the session.
+  raw_buffer.v3.RawBuffer cleartext_socket_config = 1;
```
Comment (mattklein123) [inline api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto:33]: Should this be required?

File: api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  // Configuration for TLS socket.
+  transport_sockets.tls.v3.DownstreamTlsContext tls_socket_config = 2
+      [(validate.rules).message = {required: true}];
+
+  // Maximum number of bytes allowed to be exchanged in clear-text.
+  // StartTls transport socket monitors how many bytes are
+  // exchanged in the initial clear-text phase and when that number
+  // exceeds this configuration value, the connection is terminated.
+  // Value of zero means no limit. Defaults to zero.
+  google.protobuf.UInt32Value max_cleartext_bytes = 3;
```
Comment (mattklein123) [inline api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto:44]: If the default is zero I would make this a basic uint32

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks, at a high level this LGTM. This is small enough that I'm not sure it's worth merging as is. Do you want to provide the implementation (even if without tests) and we can review the whole thing together?

/wait

File: api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto (hunk 5 above)
Comment (lizan) [reply api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto:33]: I think we can make it optional, as raw buffer doesn't have any config.

File: api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto (hunk 6 above)
Comment (lizan) [reply api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto:44]: +1

**Comment by cpakulski** [comment]
Thanks for reviewing. It makes total sense to add socket implementation to this PR.

**Comment by cpakulski** [comment]
WIP.

File: api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto (hunk 5 above)
Comment (cpakulski) [reply api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto:33]: I marked it optional in the comment. Not sure if there is "optional" keyword like this:
**optional** raw_buffer.v3.RawBuffer cleartext_socket_config = 1;

File: api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto (hunk 6 above)
Comment (cpakulski) [reply api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto]: Changed to uint32.

**Comment by cpakulski** [comment]
Originally this PR contained only api definition. Since then I added starttls socket implementation plus integration test.

The integration test simulates the intended usage of the starttls transport socket. A client opens a clear-text connection and as protocol exchange progresses, a filter understanding the protocol may instruct the transport socket to start using tls. Switching from clear-text to tls must happen without closing the socket: tls handshake starts over the same socket where clear-text exchange happened.

Testing/coverage should still be expanded, but I believe that there is enough content right now to see if this is going in the right direction.

File: api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto (hunk 5 above)
Comment (lizan) [reply api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto:33]: there's no `optional` or `required` in proto syntax, it's only about whther we put the PGV rules here.

File: source/extensions/quic_listeners/quiche/quic_filter_manager_connection_impl.h [hunk 7]
```diff
@@ -89,6 +89,11 @@ class QuicFilterManagerConnectionImpl : public Network::ConnectionImplBase {
   StreamInfo::StreamInfo& streamInfo() override { return stream_info_; }
   const StreamInfo::StreamInfo& streamInfo() const override { return stream_info_; }
   absl::string_view transportFailureReason() const override { return transport_failure_reason_; }
+  bool startSecureTransport() override {
+    RELEASE_ASSERT(false, "");
```
Comment (lizan) [inline source/extensions/quic_listeners/quiche/quic_filter_manager_connection_impl.h:93]: just do NOT_IMPLEMENTED? that will crash as well.

**Comment by mattklein123** [comment]
Will review once @lizan has approved. Thank you!

**Comment by cpakulski** [comment]
This is the final version (ready for review) which passes all CI tests.

startTls is trivial wrapper around raw_buffer and tls sockets. It starts using raw_buffer and can be modified to use tls. The crux here was how to use it from a filter. Integration tests implement a filter which instructs starttls socket to start using tls from `doRead` and `doWrite` methods. The filter must be able to send a protocol specific confirmation to a client before switching to tls. The problem was that writing operation is event driven, but call to transport socket is instantaneous. Therefore I use a callback SentBytesCb to switch to TLS after data has been transmitted to a client.

**Comment by mattklein123** [comment]
@lizan can you take a look and then pass back to me?

File: test/integration/start_tls_integration_test.cc [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+#if 0
+#endif
+
+  return Network::FilterStatus::Continue;
+}
+
+// Network::WriteFilter
+Network::FilterStatus StartTlsSwitchFilter::onWrite(Buffer::Instance&, bool) {
+  printf("Data Received onWrite!\n");
+  //data.drain(data.length());
```
Comment (fabriziomello) [inline test/integration/start_tls_integration_test.cc]: Do we need this?

File: test/integration/start_tls_integration_test.cc [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+      filter_manager.addFilter(std::make_shared<StartTlsSwitchFilter>());
+    };
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    // Using Struct instead of a custom per-filter empty config proto
+    // This is only allowed in tests.
+    return ProtobufTypes::MessagePtr{new Envoy::ProtobufWkt::Struct()};
+  }
+*/
```
Comment (fabriziomello) [inline test/integration/start_tls_integration_test.cc]: Do we need this?

File: test/integration/start_tls_integration_test.cc [hunk 10]
```diff
@@ -0,0 +1,533 @@
...
+          auto* filter_list_back = filter_chain->add_filters();
+          TestUtility::loadFromYaml(filter_yaml, *filter_list_back);
+
+          // Now move it to the front.
+          for (int i = filter_chain->filters_size() - 1; i > 0; --i) {
+            filter_chain->mutable_filters()->SwapElements(i, i - 1);
+          }
+        });
+  }
+#endif
```
Comment (fabriziomello) [inline test/integration/start_tls_integration_test.cc]: As far I understand it will never be executed, right?

File: test/integration/start_tls_integration_test.cc (hunk 8 above)
Comment (cpakulski) [reply test/integration/start_tls_integration_test.cc]: I think you are reviewing NOT the last version, as I removed those `printfs`.

File: test/integration/start_tls_integration_test.cc (hunk 9 above)
Comment (cpakulski) [reply test/integration/start_tls_integration_test.cc]: The same as above: I think you are reviewing NOT the last version.

File: test/integration/start_tls_integration_test.cc (hunk 8 above)
Comment (fabriziomello) [reply test/integration/start_tls_integration_test.cc]: Oops... sorry by the noise!!!

File: test/integration/start_tls_integration_test.cc (hunk 9 above)
Comment (fabriziomello) [reply test/integration/start_tls_integration_test.cc]: Oops... sorry by the noise!!!

File: include/envoy/network/connection.h [hunk 11]
```diff
@@ -325,6 +328,19 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
    */
   virtual absl::string_view transportFailureReason() const PURE;
 
+  /**
+   * @return std::string type of the underlying transport socket.
+   */
+  virtual std::string transportProtocol() const PURE;
```
Comment (lizan) [inline include/envoy/network/connection.h:334]: The only meaningful use of this one is to check whether transport socket can startSecureTransport, but I don't think you really need as you can call `startSecureTransport` and check its return value?

File: source/common/network/connection_impl.cc [hunk 12]
```diff
@@ -679,8 +679,15 @@ void ConnectionImpl::onWriteReady() {
...
-        cb(result.bytes_processed_);
+      auto it = bytes_sent_callbacks_.begin();
+      while (it != bytes_sent_callbacks_.end()) {
+        if ((*it)(result.bytes_processed_)) {
+          // move to the next callback.
+          it++;
+        } else {
+          // remove the current callback.
+          it = bytes_sent_callbacks_.erase(it);
+        }
```
Comment (lizan) [inline source/common/network/connection_impl.cc:690]: This change itself looks risky, can we have this part as a separate PR to get reviewed?

**Review by lizan — CHANGES_REQUESTED** [review]
Sorry for the delay, in general this is in good structure, starting from high level comments.

File: source/common/network/connection_impl.cc (hunk 12 above)
Comment (cpakulski) [reply source/common/network/connection_impl.cc:690]: Agree. Created #14245.

File: include/envoy/network/connection.h (hunk 11 above)
Comment (lizan) [reply include/envoy/network/connection.h:334]: Yeah if you want to reject at configuration level this should be part of transport socket factory but not connection.

File: source/extensions/transport_sockets/starttls/config.cc [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  auto& raw_socket_config_factory = Config::Utility::getAndCheckFactoryByName<
+      Server::Configuration::DownstreamTransportSocketConfigFactory>(
+      "envoy.transport_sockets.raw_buffer");
+
+  Network::TransportSocketFactoryPtr raw_socket_factory =
+      raw_socket_config_factory.createTransportSocketFactory(outer_config.cleartext_socket_config(),
+                                                             context, server_names);
+
+  auto& tls_socket_config_factory = Config::Utility::getAndCheckFactoryByName<
+      Server::Configuration::DownstreamTransportSocketConfigFactory>("envoy.transport_sockets.tls");
```
Comment (lizan) [inline source/extensions/transport_sockets/starttls/config.cc:32]: use `Extensions::TransportSockets::TransportSocketNames::get().Tls`, same for rawbuffer above

Comment (cpakulski) [reply source/extensions/transport_sockets/starttls/config.cc]: Done.

**Review by lizan — COMMENTED** [review]
LGTM, @mattklein123 can you take a look?

File: include/envoy/network/connection.h [hunk 14]
```diff
@@ -328,6 +328,14 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
    */
   virtual absl::string_view transportFailureReason() const PURE;
 
+  /**
+   * Instructs a transport socket to start using secure transport.
```
Comment (mattklein123) [inline include/envoy/network/connection.h:332]:
```suggestion
   * Instructs the connection to start using secure transport.
```

File: include/envoy/network/connection.h [hunk 15]
```diff
@@ -328,6 +328,14 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
...
+   * Note: Not all transport sockets support such operation.
```
Comment (mattklein123) [inline include/envoy/network/connection.h:333]:
```suggestion
   * Note: Not all underlying transport sockets support such operation.
```

File: source/common/network/raw_buffer_socket.h [hunk 16]
```diff
@@ -21,6 +21,8 @@ class RawBufferSocket : public TransportSocket, protected Logger::Loggable<Logge
   IoResult doRead(Buffer::Instance& buffer) override;
   IoResult doWrite(Buffer::Instance& buffer, bool end_stream) override;
   Ssl::ConnectionInfoConstSharedPtr ssl() const override { return nullptr; }
+  // This method must never be called for RawBufferSocket
+  bool startSecureTransport() override { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
```
Comment (mattklein123) [inline source/common/network/raw_buffer_socket.h:25]: This seems potentially a mistake that could be made? Should this should return false?

File: source/extensions/transport_sockets/alts/tsi_socket.h [hunk 17]
```diff
@@ -59,6 +59,8 @@ class TsiSocket : public Network::TransportSocket,
   absl::string_view failureReason() const override;
   bool canFlushClose() override { return handshake_complete_; }
   Envoy::Ssl::ConnectionInfoConstSharedPtr ssl() const override { return nullptr; }
+  // startSecureTransport method should not be called for this transport socket.
+  bool startSecureTransport() override { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
```
Comment (mattklein123) [inline source/extensions/transport_sockets/alts/tsi_socket.h:63]: Return false?

File: source/extensions/transport_sockets/starttls/starttls_socket.h [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+  bool startSecureTransport() override;
+
+private:
+  // Socket used in all transport socket operations.
+  // initially it is set to use raw buffer socket but
+  // can be converted to use ssl.
+  Network::TransportSocketPtr oper_socket_;
+  // Secure transport socket. It will replace raw buffer socket
+  //  when startSecureTransport is called.
+  Network::TransportSocketPtr ssl_socket_;
```
Comment (mattklein123) [inline source/extensions/transport_sockets/starttls/starttls_socket.h:60]:
```suggestion
  Network::TransportSocketPtr tls_socket_;
```

File: source/extensions/transport_sockets/starttls/starttls_socket.h [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+
+  Network::TransportSocketCallbacks* callbacks_{};
+
+  bool using_ssl_{false};
```
Comment (mattklein123) [inline source/extensions/transport_sockets/starttls/starttls_socket.h:64]:
```suggestion
  bool tls_ssl_{false};
```

File: source/extensions/transport_sockets/starttls/starttls_socket.h [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  // Method to enable TLS.
+  bool startSecureTransport() override;
+
+private:
+  // Socket used in all transport socket operations.
+  // initially it is set to use raw buffer socket but
+  // can be converted to use ssl.
+  Network::TransportSocketPtr oper_socket_;
```
Comment (mattklein123) [inline source/extensions/transport_sockets/starttls/starttls_socket.h:57]: nit: I don't know what oper_socket means. Can you name this something a bit better?

File: source/extensions/transport_sockets/starttls/starttls_socket.cc [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+namespace Extensions {
+namespace TransportSockets {
+namespace StartTls {
+
+// Switch clear-text to secure transport.
+bool StartTlsSocket::startSecureTransport() {
+  if (!using_ssl_) {
+    ssl_socket_->setTransportSocketCallbacks(*callbacks_);
+    ssl_socket_->onConnected();
+    oper_socket_ = std::move(ssl_socket_);
```
Comment (mattklein123) [inline source/extensions/transport_sockets/starttls/starttls_socket.cc:15]: Can transport sockets internally buffer? Is this guaranteed to be correct? Can we drop buffered data if there is flow control?

File: source/extensions/transport_sockets/tls/ssl_socket.cc [hunk 22]
```diff
@@ -42,6 +42,8 @@ class NotReadySslSocket : public Network::TransportSocket {
   }
   void onConnected() override {}
   Ssl::ConnectionInfoConstSharedPtr ssl() const override { return nullptr; }
+  // startSecureTransport should not be called for this transport socket.
+  bool startSecureTransport() override { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
```
Comment (mattklein123) [inline source/extensions/transport_sockets/tls/ssl_socket.cc:46]: return false

File: source/extensions/transport_sockets/tls/ssl_socket.h [hunk 23]
```diff
@@ -62,6 +62,8 @@ class SslSocket : public Network::TransportSocket,
   Network::IoResult doWrite(Buffer::Instance& write_buffer, bool end_stream) override;
   void onConnected() override;
   Ssl::ConnectionInfoConstSharedPtr ssl() const override;
+  // startSecureTransport should not be called for this transport socket.
+  bool startSecureTransport() override { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
```
Comment (mattklein123) [inline source/extensions/transport_sockets/tls/ssl_socket.h:66]: return false

File: source/extensions/transport_sockets/well_known_names.h [hunk 24]
```diff
@@ -20,6 +20,7 @@ class TransportSocketNameValues {
   const std::string Tap = "envoy.transport_sockets.tap";
   const std::string Tls = "envoy.transport_sockets.tls";
   const std::string UpstreamProxyProtocol = "envoy.transport_sockets.upstream_proxy_protocol";
+  const std::string StartTls = "envoy.transport_sockets.starttls";
```
Comment (mattklein123) [inline source/extensions/transport_sockets/well_known_names.h:23]: We are trying to get rid of these files so just inline the strings where they are needed. Same below.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks generally LGTM. Flushing out a few comments. Thank you!

/wait

File: source/extensions/transport_sockets/well_known_names.h (hunk 24 above)
Comment (cpakulski) [reply source/extensions/transport_sockets/well_known_names.h:23]: What would say if I leave it as is and create a PR which removes well_known_names.h and inlines strings instead of variables across all files?

File: source/extensions/transport_sockets/starttls/starttls_socket.cc (hunk 21 above)
Comment (cpakulski) [reply source/extensions/transport_sockets/starttls/starttls_socket.cc:15]: I think that in this case we are safe. RawBuffer::doWrite/doRead do not buffer. 
In general case, where any type of transport socket could be plugged in, it would be nice to have ::flush() method to call before switching.

File: source/extensions/transport_sockets/starttls/starttls_socket.h (hunk 20 above)
Comment (cpakulski) [reply source/extensions/transport_sockets/starttls/starttls_socket.h:57]: Renamed to `active_socket_`
 
I imagined `oper_` to mean _operating_

File: source/extensions/transport_sockets/well_known_names.h (hunk 24 above)
Comment (mattklein123) [reply source/extensions/transport_sockets/well_known_names.h:23]: OK that's fine.

File: source/extensions/transport_sockets/starttls/starttls_socket.cc (hunk 21 above)
Comment (mattklein123) [reply source/extensions/transport_sockets/starttls/starttls_socket.cc:15]: Please add a TODO/comment about this.

Comment (cpakulski) [reply source/extensions/transport_sockets/starttls/starttls_socket.cc]: Added TODO.

**Review by mattklein123 — APPROVED** [review]
Thanks!

File: test/integration/BUILD [hunk 25]
```diff
@@ -1216,6 +1216,30 @@ envoy_cc_test(
...
+    ],
+    data = [
+        "//test/config/integration/certs",
+    ],
+    deps = [
+        ":integration_lib",
+        ":starttls_integration_proto_cc_proto",
+        "//source/extensions/filters/network/tcp_proxy:config",
+        "//source/extensions/transport_sockets/raw_buffer:config",
+        "//source/extensions/transport_sockets/starttls:config",
```
Comment (antoniovicente) [inline test/integration/BUILD:1237]: We introduced a new direct dependency between core tests and //source/extensions.
Could we move this test to //test/extensions/transport_sockets ?

Comment (mattklein123) [reply test/integration/BUILD:1237]: Oops sorry I missed this. Yes @cpakulski please move.

Comment (cpakulski) [reply test/integration/BUILD:1237]: No problem - will move it.

Comment (antoniovicente) [reply test/integration/BUILD:1237]: Thanks!

Comment (cpakulski) [reply test/integration/BUILD:1237]: Moved the test: https://github.com/envoyproxy/envoy/pull/14425

File: source/common/network/raw_buffer_socket.h [hunk 26]
```diff
@@ -21,6 +21,7 @@ class RawBufferSocket : public TransportSocket, protected Logger::Loggable<Logge
   IoResult doRead(Buffer::Instance& buffer) override;
   IoResult doWrite(Buffer::Instance& buffer, bool end_stream) override;
   Ssl::ConnectionInfoConstSharedPtr ssl() const override { return nullptr; }
+  bool startSecureTransport() override { return false; }
```
Comment (antoniovicente) [inline source/common/network/raw_buffer_socket.h:24]: Test coverage for this line seems to be flaky.  Some runs seem to be hitting it but some aren't.

hit: https://storage.googleapis.com/envoy-pr/2de21bf/coverage/source/common/network/raw_buffer_socket.h.gcov.html
not hit: https://storage.googleapis.com/envoy-pr/48c0687/coverage/source/common/network/raw_buffer_socket.h.gcov.html

Comment (cpakulski) [reply source/common/network/raw_buffer_socket.h:24]: Thanks @antoniovicente. I will investigate it.
